File: sdk/vec_client.js
/**
 * VEC 2.0 Node.js Client SDK (binary frame protocol)
 *
 * VEC 2.0 is a clean break from 1.x. SDKs from 1.x are not wire-compatible.
 * See PROTOCOL-2.0.md for the full spec.
 *
 * Usage:
 *   const VecClient = require('./vec_client');
 *   const vec = new VecClient('localhost', 1920);
 *   await vec.connect();
 *
 *   // Push: vector required; data requires a label
 *   const idx = await vec.push([0.1, 0.2, 0.3]);
 *   const idx = await vec.push([0.1, 0.2, 0.3], { label: 'docs/file.pdf' });
 *   const idx = await vec.push([0.1, 0.2, 0.3], { label: 'img/cat.jpg', data: jpegBuf });
 *
 *   // Query: returns array of records {index,distance,label,data,vector}
 *   const r = await vec.query([0.1, 0.2, 0.3]);                  // default: full
 *   const r = await vec.query([0.1, 0.2, 0.3], { cosine: true });
 *   const r = await vec.query([0.1, 0.2, 0.3], { shape: VecClient.SHAPE_LABEL });
 *
 *   // QID by stored vector
 *   const r = await vec.qid(42);
 *   const r = await vec.qid('img/cat.jpg', { cosine: true });
 *
 *   // GET: single, batch, or by label
 *   const r1 = await vec.get(42);
 *   const r2 = await vec.get([0,1,2]);
 *   const r3 = await vec.get('img/cat.jpg');
 *
 *   await vec.setData(42, jpegBuf);
 *   const blob = await vec.getData(42);
 *
 *   await vec.update(42, [0.1, 0.2, 0.3]);     // vector only
 *   await vec.setLabel(42, 'img/cat.jpg');
 *   await vec.delete(42);                       // also clears label + data
 *   await vec.undo();                           // also clears last label + data
 *
 *   const { savedCount, crc } = await vec.save();
 *   const info = await vec.info();              // .protocol === 2
 *
 *   const { clusters, noise } = await vec.cluster(0.5);
 *
 * Router mode:
 *   const vec = new VecClient('localhost', 1920, 'tools');
 */
const net = require('net');

const BIN_MAGIC        = 0xF0;
const PROTOCOL_VERSION = 0x02;

const CMD_PUSH      = 0x01;
const CMD_QUERY     = 0x02;
const CMD_GET       = 0x04;
const CMD_UPDATE    = 0x06;
const CMD_DELETE    = 0x07;
const CMD_LABEL     = 0x08;
const CMD_UNDO      = 0x09;
const CMD_SAVE      = 0x0A;
const CMD_CLUSTER   = 0x0D;
const CMD_DISTINCT  = 0x0E;
const CMD_REPRESENT = 0x0F;
const CMD_INFO      = 0x10;
const CMD_QID       = 0x11;
const CMD_SET_DATA  = 0x13;
const CMD_GET_DATA  = 0x14;

const RESP_OK  = 0x00;
const RESP_ERR = 0x01;

const SHAPE_VECTOR = 0x01;
const SHAPE_LABEL  = 0x02;
const SHAPE_DATA   = 0x04;
const SHAPE_FULL   = SHAPE_VECTOR | SHAPE_LABEL | SHAPE_DATA;

const GET_MODE_SINGLE = 0x00;
const GET_MODE_BATCH  = 0x01;

const METRIC_L2     = 0x00;
const METRIC_COSINE = 0x01;

const MAX_LABEL_BYTES = 2048;
const MAX_DATA_BYTES  = 102400;

class VecClient {
    constructor(host = 'localhost', port = 1920, namespace = null) {
        this.host = host;
        this.port = port;
        this.sock = null;
        this.buffer = Buffer.alloc(0);
        this.pending = [];
        this._ns = namespace ? Buffer.from(namespace) : Buffer.alloc(0);
        this._dimCache = null;
    }

    connect() {
        return new Promise((resolve, reject) => {
            this.sock = net.createConnection(this.port, this.host, () => resolve());
            this.sock.on('error', reject);
            this.sock.on('data', (data) => {
                this.buffer = Buffer.concat([this.buffer, data]);
                this._drain();
            });
            this.sock.on('close', () => {
                while (this.pending.length > 0) {
                    const req = this.pending.shift();
                    if (req.reject) req.reject(new Error('connection closed'));
                }
            });
        });
    }

    _drain() {
        while (this.pending.length > 0) {
            const req = this.pending[0];
            if (this.buffer.length < req.n) break;
            const chunk = this.buffer.subarray(0, req.n);
            this.buffer = this.buffer.subarray(req.n);
            this.pending.shift();
            req.resolve(Buffer.from(chunk));
        }
    }

    _readBytes(n) {
        return new Promise((resolve, reject) => {
            this.pending.push({ n, resolve, reject });
            this._drain();
        });
    }

    async _recvResponse() {
        const hdr = await this._readBytes(5);
        const status = hdr[0];
        const bodyLen = hdr.readUInt32LE(1);
        const body = bodyLen > 0 ? await this._readBytes(bodyLen) : Buffer.alloc(0);
        if (status === RESP_ERR) {
            throw new Error(body.toString('utf8'));
        }
        if (status !== RESP_OK) {
            throw new Error(`unknown status byte 0x${status.toString(16)}`);
        }
        return body;
    }

    _sendFrame(cmd, label = Buffer.alloc(0), body = Buffer.alloc(0)) {
        if (label.length > MAX_LABEL_BYTES) throw new Error(`label too long (${label.length} > ${MAX_LABEL_BYTES})`);
        const head = Buffer.alloc(1 + 2 + this._ns.length + 1 + 2 + label.length + 4);
        let off = 0;
        head[off++] = BIN_MAGIC;
        head.writeUInt16LE(this._ns.length, off); off += 2;
        this._ns.copy(head, off); off += this._ns.length;
        head[off++] = cmd;
        head.writeUInt16LE(label.length, off); off += 2;
        label.copy(head, off); off += label.length;
        head.writeUInt32LE(body.length, off); off += 4;
        this.sock.write(head);
        if (body.length > 0) this.sock.write(body);
    }

    _vecToBuffer(vector) {
        const arr = new Float32Array(vector);
        return Buffer.from(arr.buffer, arr.byteOffset, arr.byteLength);
    }

    _labelBuf(label) {
        if (label === null || label === undefined) return Buffer.alloc(0);
        return Buffer.isBuffer(label) ? label : Buffer.from(String(label));
    }

    async _dim() {
        if (this._dimCache === null) this._dimCache = (await this.info()).dim;
        return this._dimCache;
    }

    _decodeRecords(body, shape, dim, withDistance) {
        if (body.length < 4) return [];
        const count = body.readUInt32LE(0);
        let off = 4;
        const out = [];
        for (let i = 0; i < count; i++) {
            const index = body.readInt32LE(off); off += 4;
            let distance = null;
            if (withDistance) { distance = body.readFloatLE(off); off += 4; }
            let label = null, data = null, vector = null;
            if (shape & SHAPE_LABEL) {
                const ll = body.readUInt32LE(off); off += 4;
                if (ll > 0) { label = body.subarray(off, off + ll).toString('utf8'); off += ll; }
            }
            if (shape & SHAPE_DATA) {
                const dl = body.readUInt32LE(off); off += 4;
                if (dl > 0) { data = Buffer.from(body.subarray(off, off + dl)); off += dl; }
            }
            if (shape & SHAPE_VECTOR) {
                const slice = body.subarray(off, off + dim * 4);
                vector = new Float32Array(new Uint8Array(slice).buffer);
                off += dim * 4;
            }
            out.push({ index, distance, label, data, vector });
        }
        return out;
    }

    // ---------------- commands ----------------

    async push(vector, opts = {}) {
        const { label = null, data = null } = opts;
        const lbl = this._labelBuf(label);
        const payload = data || Buffer.alloc(0);
        const payloadBuf = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
        if (payloadBuf.length > 0 && lbl.length === 0) throw new Error('data requires label');
        if (payloadBuf.length > MAX_DATA_BYTES) throw new Error(`data too large (${payloadBuf.length} > ${MAX_DATA_BYTES})`);
        const vecBuf = this._vecToBuffer(vector);
        const dlen = Buffer.alloc(4); dlen.writeUInt32LE(payloadBuf.length, 0);
        const body = Buffer.concat([vecBuf, dlen, payloadBuf]);
        this._sendFrame(CMD_PUSH, lbl, body);
        const resp = await this._recvResponse();
        return resp.readInt32LE(0);
    }

    async query(vector, opts = {}) {
        const { cosine = false, shape = SHAPE_FULL } = opts;
        const vecBuf = this._vecToBuffer(vector);
        const head = Buffer.from([cosine ? METRIC_COSINE : METRIC_L2, shape & 0xFF]);
        this._sendFrame(CMD_QUERY, Buffer.alloc(0), Buffer.concat([head, vecBuf]));
        const body = await this._recvResponse();
        return this._decodeRecords(body, shape, await this._dim(), true);
    }

    async qid(indexOrLabel, opts = {}) {
        const { cosine = false, shape = SHAPE_FULL } = opts;
        const head = Buffer.from([cosine ? METRIC_COSINE : METRIC_L2, shape & 0xFF]);
        if (typeof indexOrLabel === 'number') {
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(indexOrLabel, 0);
            this._sendFrame(CMD_QID, Buffer.alloc(0), Buffer.concat([head, idxBuf]));
        } else {
            this._sendFrame(CMD_QID, this._labelBuf(indexOrLabel), head);
        }
        const body = await this._recvResponse();
        return this._decodeRecords(body, shape, await this._dim(), true);
    }

    async get(target, opts = {}) {
        const { shape = SHAPE_FULL } = opts;
        if (Array.isArray(target)) {
            const head = Buffer.from([GET_MODE_BATCH, shape & 0xFF]);
            const cnt = Buffer.alloc(4); cnt.writeUInt32LE(target.length, 0);
            const idxBuf = Buffer.alloc(target.length * 4);
            for (let i = 0; i < target.length; i++) idxBuf.writeInt32LE(target[i], i * 4);
            this._sendFrame(CMD_GET, Buffer.alloc(0), Buffer.concat([head, cnt, idxBuf]));
        } else if (typeof target === 'number') {
            const head = Buffer.from([GET_MODE_SINGLE, shape & 0xFF]);
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(target, 0);
            this._sendFrame(CMD_GET, Buffer.alloc(0), Buffer.concat([head, idxBuf]));
        } else {
            const head = Buffer.from([GET_MODE_SINGLE, shape & 0xFF]);
            this._sendFrame(CMD_GET, this._labelBuf(target), head);
        }
        const body = await this._recvResponse();
        return this._decodeRecords(body, shape, await this._dim(), false);
    }

    async update(indexOrLabel, vector) {
        const vecBuf = this._vecToBuffer(vector);
        if (typeof indexOrLabel === 'number') {
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(indexOrLabel, 0);
            this._sendFrame(CMD_UPDATE, Buffer.alloc(0), Buffer.concat([idxBuf, vecBuf]));
        } else {
            this._sendFrame(CMD_UPDATE, this._labelBuf(indexOrLabel), vecBuf);
        }
        await this._recvResponse();
    }

    async setLabel(index, label) {
        const lbl = label ? this._labelBuf(label) : Buffer.alloc(0);
        const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(index, 0);
        this._sendFrame(CMD_LABEL, lbl, idxBuf);
        await this._recvResponse();
    }

    async delete(indexOrLabel) {
        if (typeof indexOrLabel === 'number') {
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(indexOrLabel, 0);
            this._sendFrame(CMD_DELETE, Buffer.alloc(0), idxBuf);
        } else {
            this._sendFrame(CMD_DELETE, this._labelBuf(indexOrLabel));
        }
        await this._recvResponse();
    }

    async undo() {
        this._sendFrame(CMD_UNDO);
        await this._recvResponse();
    }

    async save() {
        this._sendFrame(CMD_SAVE);
        const body = await this._recvResponse();
        return { savedCount: body.readUInt32LE(0), crc: body.readUInt32LE(4) };
    }

    async info() {
        this._sendFrame(CMD_INFO);
        const body = await this._recvResponse();
        let off = 0;
        const dim     = body.readInt32LE(off); off += 4;
        const count   = body.readInt32LE(off); off += 4;
        const deleted = body.readInt32LE(off); off += 4;
        const fmt     = body[off] === 1 ? 'f16' : 'f32'; off += 1;
        const mtime   = Number(body.readBigInt64LE(off)); off += 8;
        const crc     = body.readUInt32LE(off); off += 4;
        const crcOk   = body[off]; off += 1;
        const nameLen = body.readUInt32LE(off); off += 4;
        const name    = body.subarray(off, off + nameLen).toString('utf8'); off += nameLen;
        const protocol = body[off]; off += 1;
        return { dim, count, deleted, fmt, mtime, crc, crcOk, name, protocol };
    }

    async setData(indexOrLabel, data) {
        const payload = data ? (Buffer.isBuffer(data) ? data : Buffer.from(data)) : Buffer.alloc(0);
        if (payload.length > MAX_DATA_BYTES) throw new Error(`data too large (${payload.length} > ${MAX_DATA_BYTES})`);
        const dlen = Buffer.alloc(4); dlen.writeUInt32LE(payload.length, 0);
        if (typeof indexOrLabel === 'number') {
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(indexOrLabel, 0);
            this._sendFrame(CMD_SET_DATA, Buffer.alloc(0), Buffer.concat([idxBuf, dlen, payload]));
        } else {
            this._sendFrame(CMD_SET_DATA, this._labelBuf(indexOrLabel), Buffer.concat([dlen, payload]));
        }
        await this._recvResponse();
    }

    async getData(indexOrLabel) {
        if (typeof indexOrLabel === 'number') {
            const idxBuf = Buffer.alloc(4); idxBuf.writeInt32LE(indexOrLabel, 0);
            this._sendFrame(CMD_GET_DATA, Buffer.alloc(0), idxBuf);
        } else {
            this._sendFrame(CMD_GET_DATA, this._labelBuf(indexOrLabel));
        }
        const body = await this._recvResponse();
        const dlen = body.readUInt32LE(0);
        return Buffer.from(body.subarray(4, 4 + dlen));
    }

    async cluster(eps, opts = {}) {
        const { cosine = false, minPts = 2 } = opts;
        const data = Buffer.alloc(9);
        data.writeFloatLE(eps, 0);
        data[4] = cosine ? METRIC_COSINE : METRIC_L2;
        data.writeInt32LE(minPts, 5);
        this._sendFrame(CMD_CLUSTER, Buffer.alloc(0), data);
        const text = (await this._recvResponse()).toString('utf8');
        const lines = [];
        for (const raw of text.split('\n')) {
            const line = raw.trim();
            if (!line || line === 'end') continue;
            lines.push(line.split(',').filter(m => m));
        }
        const noise = lines.length > 0 ? lines.pop() : [];
        return { clusters: lines, noise };
    }

    async distinct(k, opts = {}) {
        const { cosine = false } = opts;
        const data = Buffer.alloc(5);
        data.writeInt32LE(k, 0);
        data[4] = cosine ? METRIC_COSINE : METRIC_L2;
        this._sendFrame(CMD_DISTINCT, Buffer.alloc(0), data);
        return this._parseIndexLines(await this._recvResponse());
    }

    async represent(eps, opts = {}) {
        const { cosine = false, minPts = 2 } = opts;
        const data = Buffer.alloc(9);
        data.writeFloatLE(eps, 0);
        data[4] = cosine ? METRIC_COSINE : METRIC_L2;
        data.writeInt32LE(minPts, 5);
        this._sendFrame(CMD_REPRESENT, Buffer.alloc(0), data);
        return this._parseIndexLines(await this._recvResponse());
    }

    _parseIndexLines(body) {
        const out = [];
        for (const raw of body.toString('utf8').split('\n')) {
            const line = raw.trim();
            if (!line || line === 'end') continue;
            const n = parseInt(line, 10);
            out.push(!isNaN(n) && n.toString() === line ? n : line);
        }
        return out;
    }

    close() {
        if (this.sock) this.sock.destroy();
    }
}

VecClient.SHAPE_VECTOR = SHAPE_VECTOR;
VecClient.SHAPE_LABEL  = SHAPE_LABEL;
VecClient.SHAPE_DATA   = SHAPE_DATA;
VecClient.SHAPE_FULL   = SHAPE_FULL;
VecClient.PROTOCOL_VERSION = PROTOCOL_VERSION;

module.exports = VecClient;

File: sdk/vec_client.py
"""
VEC 2.0 Python Client SDK (binary frame protocol)

VEC 2.0 is a clean break from 1.x. SDKs from 1.x are not wire-compatible.
See PROTOCOL-2.0.md for the full spec.

Usage:
    from vec_client import VecClient, SHAPE_VECTOR, SHAPE_LABEL, SHAPE_DATA, SHAPE_FULL

    vec = VecClient("localhost", 1920)

    # Push: vector required; data requires a label
    idx = vec.push([0.1, 0.2, 0.3])
    idx = vec.push([0.1, 0.2, 0.3], label="docs/file.pdf")
    idx = vec.push([0.1, 0.2, 0.3], label="img/cat.jpg", data=open("cat.jpg","rb").read())

    # Query: returns list of VecResult with optional label/data/vector per shape mask
    results = vec.query([0.1, 0.2, 0.3])                       # default: full record
    results = vec.query([0.1, 0.2, 0.3], cosine=True)
    results = vec.query([0.1, 0.2, 0.3], shape=SHAPE_LABEL)    # lean: index+label only

    # QID: query using a stored vector by index or label
    results = vec.qid(42)
    results = vec.qid("img/cat.jpg", cosine=True)

    # GET: single or batch fetch
    rec  = vec.get(42)                          # single, full record
    recs = vec.get([0, 1, 2])                   # batch
    recs = vec.get("img/cat.jpg")               # by label (may return multiple)

    # Sidecar payload management
    vec.set_data(42, b"...binary blob...")      # ≤100KB
    blob = vec.get_data(42)

    vec.update(42, [0.1, 0.2, 0.3])             # vector only; use set_label / set_data for the rest
    vec.set_label(42, "img/cat.jpg")
    vec.delete(42)                              # also clears label and data
    vec.undo()                                  # also clears last label and data

    saved, crc = vec.save()
    info = vec.info()                           # includes 'protocol' (= 2)

    # Clustering still works; CLUSTER body is legacy text inside binary envelope
    clusters, noise = vec.cluster(0.5)

Router mode:
    vec = VecClient("localhost", 1920, namespace="tools")
"""
import socket
import struct
import numpy as np

BIN_MAGIC        = 0xF0
PROTOCOL_VERSION = 0x02

CMD_PUSH      = 0x01
CMD_QUERY     = 0x02
CMD_GET       = 0x04
CMD_UPDATE    = 0x06
CMD_DELETE    = 0x07
CMD_LABEL     = 0x08
CMD_UNDO      = 0x09
CMD_SAVE      = 0x0A
CMD_CLUSTER   = 0x0D
CMD_DISTINCT  = 0x0E
CMD_REPRESENT = 0x0F
CMD_INFO      = 0x10
CMD_QID       = 0x11
CMD_SET_DATA  = 0x13
CMD_GET_DATA  = 0x14

RESP_OK  = 0x00
RESP_ERR = 0x01

SHAPE_VECTOR = 0x01
SHAPE_LABEL  = 0x02
SHAPE_DATA   = 0x04
SHAPE_FULL   = SHAPE_VECTOR | SHAPE_LABEL | SHAPE_DATA  # 0x07 — default

GET_MODE_SINGLE = 0x00
GET_MODE_BATCH  = 0x01

METRIC_L2     = 0x00
METRIC_COSINE = 0x01

MAX_LABEL_BYTES = 2048
MAX_DATA_BYTES  = 102400


class VecError(RuntimeError):
    pass


class VecRecord:
    """One result record from query / qid / get."""
    __slots__ = ("index", "distance", "label", "data", "vector")

    def __init__(self, index, distance=None, label=None, data=None, vector=None):
        self.index    = index
        self.distance = distance
        self.label    = label
        self.data     = data
        self.vector   = vector

    def __repr__(self):
        parts = [f"index={self.index}"]
        if self.distance is not None: parts.append(f"distance={self.distance:.6f}")
        if self.label is not None:    parts.append(f"label={self.label!r}")
        if self.data is not None:     parts.append(f"data=<{len(self.data)} bytes>")
        if self.vector is not None:   parts.append(f"vector=<{len(self.vector)} f32>")
        return "VecRecord(" + ", ".join(parts) + ")"


class VecClient:
    def __init__(self, host="localhost", port=1920, namespace=None):
        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.sock.connect((host, port))
        self._ns = namespace.encode() if namespace else b""
        self._dim_cache = None

    # ---------- low-level wire helpers ----------

    def _recv_exact(self, n):
        buf = bytearray()
        while len(buf) < n:
            chunk = self.sock.recv(n - len(buf))
            if not chunk:
                raise VecError("connection closed")
            buf += chunk
        return bytes(buf)

    def _send_frame(self, cmd, label=b"", body=b""):
        if len(label) > MAX_LABEL_BYTES:
            raise VecError(f"label too long ({len(label)} > {MAX_LABEL_BYTES})")
        hdr = (
            bytes([BIN_MAGIC])
            + struct.pack("<H", len(self._ns)) + self._ns
            + bytes([cmd])
            + struct.pack("<H", len(label)) + label
            + struct.pack("<I", len(body))
        )
        self.sock.sendall(hdr + body)

    def _recv_response(self):
        """Read a 2.0 response envelope. Returns body bytes on OK, raises VecError on ERR."""
        hdr = self._recv_exact(5)
        status = hdr[0]
        body_len = struct.unpack_from("<I", hdr, 1)[0]
        body = self._recv_exact(body_len) if body_len > 0 else b""
        if status == RESP_ERR:
            raise VecError(body.decode("utf-8", errors="replace"))
        if status != RESP_OK:
            raise VecError(f"unknown status byte 0x{status:02x}")
        return body

    @staticmethod
    def _label_bytes(label):
        if label is None:
            return b""
        if isinstance(label, str):
            return label.encode()
        return bytes(label)

    def _dim(self):
        if self._dim_cache is None:
            self._dim_cache = self.info()["dim"]
        return self._dim_cache

    # ---------- record decoding (shape-aware) ----------

    @staticmethod
    def _decode_records(body, shape, dim, with_distance):
        if len(body) < 4:
            return []
        count = struct.unpack_from("<I", body, 0)[0]
        off = 4
        out = []
        for _ in range(count):
            (idx,) = struct.unpack_from("<i", body, off); off += 4
            dist = None
            if with_distance:
                (dist,) = struct.unpack_from("<f", body, off); off += 4
            label = None
            data  = None
            vector = None
            if shape & SHAPE_LABEL:
                (ll,) = struct.unpack_from("<I", body, off); off += 4
                if ll > 0:
                    label = body[off:off+ll].decode("utf-8", errors="replace")
                    off += ll
            if shape & SHAPE_DATA:
                (dl,) = struct.unpack_from("<I", body, off); off += 4
                if dl > 0:
                    data = bytes(body[off:off+dl])
                    off += dl
            if shape & SHAPE_VECTOR:
                vector = np.frombuffer(body, dtype=np.float32, count=dim, offset=off).copy()
                off += dim * 4
            out.append(VecRecord(idx, distance=dist, label=label, data=data, vector=vector))
        return out

    # ---------- commands ----------

    def push(self, vector, label=None, data=None):
        """Push a vector. Returns slot index. data requires label."""
        arr = np.asarray(vector, dtype=np.float32)
        lbl = self._label_bytes(label)
        payload = data if data is not None else b""
        if len(payload) > 0 and not lbl:
            raise VecError("data requires label")
        if len(payload) > MAX_DATA_BYTES:
            raise VecError(f"data too large ({len(payload)} > {MAX_DATA_BYTES})")
        body = arr.tobytes() + struct.pack("<I", len(payload)) + payload
        self._send_frame(CMD_PUSH, label=lbl, body=body)
        body = self._recv_response()
        return struct.unpack("<i", body)[0]

    def query(self, vector, cosine=False, shape=SHAPE_FULL):
        """L2 (or cosine) nearest-neighbor search. Returns list of VecRecord."""
        arr = np.asarray(vector, dtype=np.float32)
        body = bytes([METRIC_COSINE if cosine else METRIC_L2, shape & 0xFF]) + arr.tobytes()
        self._send_frame(CMD_QUERY, body=body)
        return self._decode_records(self._recv_response(), shape, self._dim(), with_distance=True)

    def qid(self, index_or_label, cosine=False, shape=SHAPE_FULL):
        """Query using a stored vector as the source. Accepts int index or str label."""
        metric = METRIC_COSINE if cosine else METRIC_L2
        if isinstance(index_or_label, int):
            body = bytes([metric, shape & 0xFF]) + struct.pack("<i", index_or_label)
            self._send_frame(CMD_QID, body=body)
        else:
            lbl = self._label_bytes(index_or_label)
            body = bytes([metric, shape & 0xFF])
            self._send_frame(CMD_QID, label=lbl, body=body)
        return self._decode_records(self._recv_response(), shape, self._dim(), with_distance=True)

    def get(self, target, shape=SHAPE_FULL):
        """Fetch records.

        target: int (single by index) | str (single by label, may return multiple) |
                list[int] (batch by indices).
        Returns list of VecRecord.
        """
        if isinstance(target, list):
            indices = target
            body = (bytes([GET_MODE_BATCH, shape & 0xFF])
                    + struct.pack("<I", len(indices))
                    + struct.pack(f"<{len(indices)}i", *indices))
            self._send_frame(CMD_GET, body=body)
        elif isinstance(target, int):
            body = bytes([GET_MODE_SINGLE, shape & 0xFF]) + struct.pack("<i", target)
            self._send_frame(CMD_GET, body=body)
        else:
            lbl = self._label_bytes(target)
            body = bytes([GET_MODE_SINGLE, shape & 0xFF])
            self._send_frame(CMD_GET, label=lbl, body=body)
        return self._decode_records(self._recv_response(), shape, self._dim(), with_distance=False)

    def update(self, index_or_label, vector):
        """Overwrite vector in-place. Does NOT touch label or data."""
        arr = np.asarray(vector, dtype=np.float32)
        if isinstance(index_or_label, int):
            body = struct.pack("<i", index_or_label) + arr.tobytes()
            self._send_frame(CMD_UPDATE, body=body)
        else:
            self._send_frame(CMD_UPDATE,
                             label=self._label_bytes(index_or_label),
                             body=arr.tobytes())
        self._recv_response()

    def set_label(self, index, label):
        """Set or clear label for a slot. label=None or empty clears."""
        lbl = self._label_bytes(label) if label else b""
        self._send_frame(CMD_LABEL, label=lbl, body=struct.pack("<i", index))
        self._recv_response()

    def delete(self, index_or_label):
        """Tombstone a slot. Also clears its label and data."""
        if isinstance(index_or_label, int):
            self._send_frame(CMD_DELETE, body=struct.pack("<i", index_or_label))
        else:
            self._send_frame(CMD_DELETE, label=self._label_bytes(index_or_label))
        self._recv_response()

    def undo(self):
        """Remove last pushed record. Also clears its label and data."""
        self._send_frame(CMD_UNDO)
        self._recv_response()

    def save(self):
        """Flush to disk. Returns (saved_count, crc32)."""
        self._send_frame(CMD_SAVE)
        body = self._recv_response()
        return struct.unpack("<II", body)

    def info(self):
        """DB metadata: dim, count, deleted, fmt, mtime, crc, crc_ok, name, protocol."""
        self._send_frame(CMD_INFO)
        body = self._recv_response()
        off = 0
        dim, count, deleted = struct.unpack_from("<iii", body, off); off += 12
        fmt    = body[off]; off += 1
        mtime, = struct.unpack_from("<q", body, off); off += 8
        crc,   = struct.unpack_from("<I", body, off); off += 4
        crc_ok = body[off]; off += 1
        name_len, = struct.unpack_from("<I", body, off); off += 4
        name = body[off:off+name_len].decode("utf-8", errors="replace"); off += name_len
        protocol = body[off]; off += 1
        return {
            "dim": dim, "count": count, "deleted": deleted,
            "fmt": "f16" if fmt == 1 else "f32",
            "mtime": mtime, "crc": crc, "crc_ok": crc_ok,
            "name": name, "protocol": protocol,
        }

    def set_data(self, index_or_label, data):
        """Set or clear sidecar payload. data=None or b"" clears."""
        payload = data or b""
        if len(payload) > MAX_DATA_BYTES:
            raise VecError(f"data too large ({len(payload)} > {MAX_DATA_BYTES})")
        if isinstance(index_or_label, int):
            body = struct.pack("<iI", index_or_label, len(payload)) + payload
            self._send_frame(CMD_SET_DATA, body=body)
        else:
            body = struct.pack("<I", len(payload)) + payload
            self._send_frame(CMD_SET_DATA, label=self._label_bytes(index_or_label), body=body)
        self._recv_response()

    def get_data(self, index_or_label):
        """Fetch sidecar payload. Returns bytes (empty if no data)."""
        if isinstance(index_or_label, int):
            self._send_frame(CMD_GET_DATA, body=struct.pack("<i", index_or_label))
        else:
            self._send_frame(CMD_GET_DATA, label=self._label_bytes(index_or_label))
        body = self._recv_response()
        (dlen,) = struct.unpack_from("<I", body, 0)
        return bytes(body[4:4+dlen])

    # ---------- cluster / distinct / represent ----------
    # Body of these responses is still the legacy line-based text from 1.x,
    # wrapped in the 2.0 binary envelope. Decoded here for convenience.

    def cluster(self, eps, cosine=False, min_pts=2):
        """DBSCAN. Returns (clusters, noise). Each is list[list[str]] of label-or-index strings."""
        body = struct.pack("<fBi", float(eps), 1 if cosine else 0, int(min_pts))
        self._send_frame(CMD_CLUSTER, body=body)
        text = self._recv_response().decode("utf-8", errors="replace")
        return self._parse_text_lines_with_end(text)

    def distinct(self, k, cosine=False):
        """Farthest-point sampling. CPU build returns 'not available' error."""
        body = struct.pack("<iB", int(k), 1 if cosine else 0)
        self._send_frame(CMD_DISTINCT, body=body)
        text = self._recv_response().decode("utf-8", errors="replace")
        return self._parse_text_index_lines(text)

    def represent(self, eps, cosine=False, min_pts=2):
        """One representative per cluster. CPU build returns 'not available' error."""
        body = struct.pack("<fBi", float(eps), 1 if cosine else 0, int(min_pts))
        self._send_frame(CMD_REPRESENT, body=body)
        text = self._recv_response().decode("utf-8", errors="replace")
        return self._parse_text_index_lines(text)

    @staticmethod
    def _parse_text_lines_with_end(text):
        lines = []
        for line in text.split("\n"):
            line = line.strip()
            if not line or line == "end":
                continue
            lines.append([m for m in line.split(",") if m])
        if lines:
            return lines[:-1], lines[-1]
        return [], []

    @staticmethod
    def _parse_text_index_lines(text):
        out = []
        for line in text.split("\n"):
            line = line.strip()
            if not line or line == "end":
                continue
            try:
                out.append(int(line))
            except ValueError:
                out.append(line)
        return out

    # ---------- lifecycle ----------

    def close(self):
        try:
            self.sock.close()
        except Exception:
            pass

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()
